src/dbnode/encoding/istream.go

```diff
@@ -21,169 +21,119 @@
 package encoding
 
 import (
-	"bufio"
 	"io"
-	"math"
-)
 
-// istream encapsulates a readable stream.
-type istream struct {
-	r         *bufio.Reader // encoded stream
-	err       error         // error encountered
-	current   byte          // current byte we are working off of
-	buffer    []byte        // buffer for reading in multiple bytes
-	remaining uint          // bits remaining in current to be read
-}
+	"github.com/m3db/m3/src/dbnode/x/xio"
+)
 
-// NewIStream creates a new Istream
-func NewIStream(reader io.Reader, bufioSize int) IStream {
-	return &istream{
-		r: bufio.NewReaderSize(reader, bufioSize),
-		// Buffer meant to hold uint64 size of bytes.
-		buffer: make([]byte, 8),
-	}
+// IStream encapsulates a readable stream.
+type IStream struct {
+	r         xio.Reader64
+	err       error  // error encountered
+	current   uint64 // current uint64 we are working off of
+	index     int    // current index within data slice
+	remaining uint8  // bits remaining in current to be read
 }
 
-func (is *istream) ReadBit() (Bit, error) {
-	if is.err != nil {
-		return 0, is.err
-	}
-	if is.remaining == 0 {
-		if err := is.readByteFromStream(); err != nil {
-			return 0, err
-		}
-	}
-	return Bit(is.consumeBuffer(1)), nil
+// NewIStream creates a new IStream
+func NewIStream(reader64 xio.Reader64) *IStream {
```

Comment:
A good question. `IStream` used to be an interface before this change, and now it has become a struct. I think I left it returning a pointer in order to avoid accidental semantics change at the call sites (to avoid any potential pass-by-value of this struct down the stream).